server/src/main/java/org/apache/iotdb/db/engine/merge/task/RecoverMergeTask.java

```diff
@@ -80,7 +95,57 @@ public void recoverMerge() throws IOException, MetadataException {
     }
   }
 
-  private void resumeAfterFilesLogged() throws IOException {
+  private void handleWhenAllSourceFilesExist() throws IOException {
     cleanUp(false);
+    appendNewModificationToOldModsFile(resource.getSeqFiles());
+    appendNewModificationToOldModsFile(resource.getUnseqFiles());
+    tsFileManagement.removeMergingModification();
+  }
+
+  private void handleWhenSomeSourceFilesLost() throws IOException {
+    for (TsFileResource sourceSeqResource : resource.getSeqFiles()) {
+      File targetFile = modifyTsFileNameUnseqMergCnt(sourceSeqResource.getTsFile());
+      File targetFileResource = new File(targetFile.getPath() + TsFileResource.RESOURCE_SUFFIX);
+      File tmpTargetFile = new File(sourceSeqResource.getTsFilePath() + MERGE_SUFFIX);
+      // move to target file and target resource file
+      if (!targetFileResource.exists()) {
+        if (!targetFile.exists()) {
+          // move target file
+          FSFactoryProducer.getFSFactory().moveFile(tmpTargetFile, targetFile);
+        }
+        // move target resource file
+        FSFactoryProducer.getFSFactory()
```

Comment:
If the target file already exists, its tsfileResource must exist, because SGP recovery generates a tsfileResource for a tsfile that does not have a tsfileResource (see `TsFileRecoverPerformer::recover`). If the target file does not exist and the corresponding temporary file does, the tsfileResource should be regenerated by reading from the temporary file rather than moving the tsfileResource of source file directly.

Reply:
resolved